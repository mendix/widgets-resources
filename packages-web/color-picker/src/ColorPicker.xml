<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.colorpicker.ColorPicker" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/">
    <name>Color picker</name>
    <description>Pick a color from color input</description>
    <icon/>
    <properties>
        <property key="colorAttribute" type="attribute">
            <caption>Color attribute</caption>
            <category>Data source</category>
            <description>Attribute containing a valid color: #d0d0d0, rgb(115,159,159) or rgba(195,226,226,1). Non-code color values such as 'red' are not supported</description>
            <attributeTypes>
                <attributeType name="String"/>
            </attributeTypes>
        </property>
        <property key="format" type="enumeration" defaultValue="hex">
            <caption>Color format</caption>
            <category>Data source</category>
            <description>The output formatting of the selected color</description>
            <enumerationValues>
                <enumerationValue key="hex">hex</enumerationValue>
                <enumerationValue key="rgb">rgb</enumerationValue>
                <enumerationValue key="rgba">rgba</enumerationValue>
            </enumerationValues>
        </property>
        <property key="editable" type="enumeration" defaultValue="default">
            <caption>Editable</caption>
            <category>Data source</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="default">Default</enumerationValue>
                <enumerationValue key="never">Never</enumerationValue>
            </enumerationValues>
        </property>
        <property key="defaultColors" type="object" isList="true" required="false">
            <caption>Default colors</caption>
            <category>Data source</category>
            <description>The list of predefined colors. This can only be used in combination with 'Block', 'Sketch', 'Circle', 'Compact' or 'Twitter'</description>
            <properties>
                <property key="color" type="string">
                    <caption>Color</caption>
                    <category>Data source</category>
                    <description>Valid color value: #d0d0d0, rgb(115,159,159) or rgba(195,226,226,1)</description>
                </property>
            </properties>
        </property>
        <property key="mode" type="enumeration" defaultValue="input">
            <caption>Mode</caption>
            <category>Appearance</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="input">Input box</enumerationValue>
                <enumerationValue key="popover">Button</enumerationValue>
                <enumerationValue key="inline">Inline</enumerationValue>
            </enumerationValues>
        </property>
        <property key="type" type="enumeration" defaultValue="sketch">
            <caption>Picker type</caption>
            <category>Appearance</category>
            <description>For samples see: http://casesandberg.github.io/react-color/</description>
            <enumerationValues>
                <enumerationValue key="sketch">Sketch</enumerationValue>
                <enumerationValue key="chrome">Chrome</enumerationValue>
                <enumerationValue key="block">Block</enumerationValue>
                <enumerationValue key="github">Github</enumerationValue>
                <enumerationValue key="twitter">Twitter</enumerationValue>
                <enumerationValue key="circle">Circle</enumerationValue>
                <enumerationValue key="hue">Hue</enumerationValue>
                <enumerationValue key="slider">Slider</enumerationValue>
                <enumerationValue key="compact">Compact</enumerationValue>
                <enumerationValue key="material">Material</enumerationValue>
                <enumerationValue key="swatches">Swatches</enumerationValue>
            </enumerationValues>
        </property>
        <property key="label" type="translatableString" required="false">
            <caption>Label</caption>
            <category>Appearance</category>
            <description/>
            <translations>
                <translation lang="en_US">Label</translation>
                <translation lang="nl_NL">Label</translation>
            </translations>
        </property>
        <property key="labelWidth" type="integer" defaultValue="3">
            <caption>Label width</caption>
            <category>Appearance</category>
            <description>The weight of the label, which must be a value between 1 and 11</description>
        </property>
        <property key="onChangeEvent" type="enumeration" defaultValue="doNothing">
            <caption>On change</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="onChangeMicroflow" type="microflow" required="false">
            <caption>Microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on color change</description>
            <returnType type="Void"/>
        </property>
        <property key="onChangeNanoflow" type="nanoflow" required="false">
            <caption>Nanoflow</caption>
            <category>Events</category>
            <description>The nanoflow that will be executed on color change</description>
            <returnType type="Void"/>
        </property>
    </properties>
</widget>
