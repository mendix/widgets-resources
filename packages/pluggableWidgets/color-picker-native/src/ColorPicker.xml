<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.native.colorpicker.ColorPicker" supportedPlatform="Native" needsEntityContext="true"
        offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Color picker</name>
    <description>Pick a color from a circular color input</description>
    <icon>
        iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAMAAACdt4HsAAAA5FBMVEUAAACAgIBVVVVgYGBVVVVmZmZdXV1iYmJVVVVeXl5ZWVlVVVVVVVVYWFhJtv9bW1tXV1dMs/pVVVVYWFhJs/hXV1dLtPhWVlZJsvhXV1dWVlZXV1dXV1dVVVVXV1dWVlZWVlZWVlZWVlZVVVVVVVVWVlZIsfdJsfdJsPdJsPhWVlZWVlZVVVVVVVVWVlZVVVVVVVVWVlZWVlZVVVVVVVVVVVVVVVVJsPdVVVVVVVVWVlZJsPdVVVVWVlZIsPdJsfdVVVVIsPdWVlZVVVVJsfhVVVVWVlZVVVVWVlZVVVVIsPdVVVWIl4c0AAAASnRSTlMAAgMICQoLDRITFBUYGhwcLzY2Q0ZGR0dJSUpPVWNkgoOFkpmcoaanqKyytba5vsXGx9bY293e4ODh4uPj5ebo6u3u7/Lz9Pb9/n9E/agAAAExSURBVHgB7dPZTsJAFMbxT0CpWivgIgiKslRxQUQEpbWAssic938fD4wx3KDOaWJC7O/q3Hz/pGkGke+lSs3RqFlKQSZemdLctBKHQKJBXxoJmKvSgiqMZYi1C7ZdaBPLwFSZV60kWLLFZxmmZqu8PvOzFkxNeGXp0+JzAlPEFu+/DUSBnE9L+Dn8hk9L+eED4T9h5X5j1vOyoQKeUl6ogGL/OxAFokCfR7Z+TDaffZiq8+pUP+cTPuswVeRVdxPMeuazCFNbY57d7wI7d3yMt2HsjNjbwd4rsXOYi10Re3wgdh2DwPoNfbrdgMw7aWtY2cCT3r9A6qhLzD+G3GGns48QnKFSQwdyNcVqEEuruTSkLnXgAlI9HehBKtCBAFKuDriQcgaKDRyIOW4QuD/tIx9N6qiDmf0nQgAAAABJRU5ErkJggg==
    </icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="color" type="attribute">
                    <caption>Color</caption>
                    <description>Attribute containing a valid color like #d0d0d0, red, rgba(255,255,255,0).</description>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
                <property key="format" type="enumeration" defaultValue="hex">
                    <caption>Color format</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="hex">Hexadecimal</enumerationValue>
                        <enumerationValue key="hsl">HSL</enumerationValue>
                        <enumerationValue key="hsv">HSV</enumerationValue>
                        <enumerationValue key="rgb">RGB</enumerationValue>
                    </enumerationValues>
                </property>
                <systemProperty key="Editability" />
            </propertyGroup>
            <propertyGroup caption="Show">
                <property key="showPreview" type="boolean" defaultValue="true">
                    <caption>Preview</caption>
                    <description />
                </property>
                <property key="showSaturation" type="boolean" defaultValue="true">
                    <caption>Saturation</caption>
                    <description>Enables a slider to change the intensity of the color.</description>
                </property>
                <property key="showLightness" type="boolean" defaultValue="true">
                    <caption>Brightness</caption>
                    <description>Enables a slider to make the color darker or lighter.</description>
                </property>
                <property key="showAlpha" type="boolean" defaultValue="false">
                    <caption>Transparency</caption>
                    <description>Enables a slider to make the color more transparent or opaque.</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onChange" type="action" required="false">
                    <caption>On change</caption>
                    <description>Called when the user releases the handle and the value has changed.</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Visibility" />
                <systemProperty key="Name"/>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
