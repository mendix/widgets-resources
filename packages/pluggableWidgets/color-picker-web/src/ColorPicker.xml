<?xml version="1.0" encoding="utf-8" ?>
<widget
    id="com.mendix.widget.custom.colorpicker.ColorPicker"
    needsEntityContext="true"
    pluginWidget="true"
    offlineCapable="true"
    xmlns="http://www.mendix.com/widget/1.0/"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd"
>
    <name>Color picker</name>
    <description>Pick a color from color input</description>
    <studioProCategory>Display</studioProCategory>
    <studioCategory>Display</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="colorAttribute" type="attribute">
                    <caption>Color attribute</caption>
                    <description>Attribute containing a valid color like #d0d0d0, rgb(115,159,159) or rgba(195,226,226,1). Non-code color values such as 'red' are not supported
                    </description>
                    <attributeTypes>
                        <attributeType name="String" />
                    </attributeTypes>
                </property>
                <property key="format" type="enumeration" defaultValue="hex">
                    <caption>Color format</caption>
                    <description>The output formatting of the selected color</description>
                    <enumerationValues>
                        <enumerationValue key="hex">HEX</enumerationValue>
                        <enumerationValue key="rgb">RGB</enumerationValue>
                        <enumerationValue key="rgba">RBGA</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Label">
                <systemProperty key="Label" />
            </propertyGroup>
            <propertyGroup caption="Editability">
                <systemProperty key="Editability" />
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Appearance">
            <property key="mode" type="enumeration" defaultValue="input">
                <caption>Mode</caption>
                <description />
                <enumerationValues>
                    <enumerationValue key="input">Input box</enumerationValue>
                    <enumerationValue key="popover">Button</enumerationValue>
                    <enumerationValue key="inline">Inline</enumerationValue>
                </enumerationValues>
            </property>
            <property key="type" type="enumeration" defaultValue="sketch">
                <caption>Picker type</caption>
                <description />
                <enumerationValues>
                    <enumerationValue key="sketch">Sketch</enumerationValue>
                    <enumerationValue key="chrome">Chrome</enumerationValue>
                    <enumerationValue key="block">Block</enumerationValue>
                    <enumerationValue key="github">Github</enumerationValue>
                    <enumerationValue key="twitter">Twitter</enumerationValue>
                    <enumerationValue key="circle">Circle</enumerationValue>
                    <enumerationValue key="hue">Hue</enumerationValue>
                    <enumerationValue key="slider">Slider</enumerationValue>
                    <enumerationValue key="compact">Compact</enumerationValue>
                    <enumerationValue key="material">Material</enumerationValue>
                    <enumerationValue key="swatches">Swatches</enumerationValue>
                </enumerationValues>
            </property>
            <property key="defaultColors" type="object" isList="true" required="false">
                <caption>Default colors</caption>
                <description>The list of predefined colors. This can only be used with picker type: 'Block', 'Sketch', 'Circle', 'Compact' or 'Twitter'
                </description>
                <properties>
                    <property key="color" type="string">
                        <caption>Color</caption>
                        <category>Data source</category>
                        <description>Valid color value: #d0d0d0, rgb(115,159,159) or rgba(195,226,226,1)</description>
                    </property>
                </properties>
            </property>
        </propertyGroup>
        <propertyGroup caption="Validation">
            <property key="invalidFormatMessage" type="textTemplate" required="false">
                <caption>Invalid format message</caption>
                <description>Message shown when the user provides a wrong input, :colors: will be replaced by a sample format
                </description>
                <translations>
                    <translation lang="en_US">Invalid format, color value should be of format like :colors:
                    </translation>
                    <translation lang="nl_NL">De opgeven kleur code is incorrect, voorbeeld :colors:</translation>
                </translations>
            </property>
        </propertyGroup>

        <propertyGroup caption="Events">
            <property key="onChange" type="action">
                <caption>On change</caption>
                <description />
            </property>
        </propertyGroup>
    </properties>
</widget>
