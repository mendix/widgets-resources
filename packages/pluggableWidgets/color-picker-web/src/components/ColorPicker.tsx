import { createElement, ReactElement, useEffect, useState, useCallback } from "react";
import { Alert } from "@mendix/piw-utils-internal/components/web";
import { ModeEnum, TypeEnum, DefaultColorsType, FormatEnum } from "../../typings/ColorPickerProps";
import {
    ColorState,
    SketchPickerProps,
    ChromePickerProps,
    TwitterPickerProps,
    HuePickerProps,
    BlockPickerProps,
    GithubPickerProps,
    CirclePickerProps,
    SwatchesPickerProps,
    SliderPickerProps,
    MaterialPickerProps,
    CompactPickerProps
} from "react-color";
import classNames from "classnames";
import { getColorPicker, parseColor, validateColorFormat, validateProps } from "../utils";
import { Input } from "./Input";
import { Button } from "./Button";

export interface ColorPickerProps {
    id: string;
    name: string;
    tabIndex?: number | undefined;
    onChange: () => void;
    onColorChange: (value: string) => void;
    mode: ModeEnum;
    type: TypeEnum;
    color: string | undefined;
    defaultColors: DefaultColorsType[];
    format: FormatEnum;
    disabled: boolean;
    invalidFormatMessage?: string | undefined;
}

export type ColorPickerConfigProps =
    | SketchPickerProps
    | ChromePickerProps
    | TwitterPickerProps
    | HuePickerProps
    | BlockPickerProps
    | GithubPickerProps
    | CirclePickerProps
    | SwatchesPickerProps
    | SliderPickerProps
    | MaterialPickerProps
    | CompactPickerProps;
export const ColorPicker = (props: ColorPickerProps): ReactElement => {
    const defaultColor = {
        hex: "#FFFFFF",
        rgb: "rgb(255,255,255)",
        rgba: "rgb(255,255,255,1)"
    };
    const { type, mode, disabled, defaultColors, color, format, invalidFormatMessage } = props;
    const colorElement = getColorPicker(type);
    const [hidden, setHidden] = useState(mode !== "inline");
    const [currentColor, setCurrentColor] = useState<string | undefined>(color);
    const [alertMessage, setAlertMessage] = useState<string | undefined>();

    const submitColor = (color: string): void => {
        setCurrentColor(color);
        props.onColorChange(color);
    };

    const validateColor = (colorValue: string): void => {
        const message = validateColorFormat(colorValue, format);
        const validProps = validateProps(props);
        const alertMessage = message ? invalidFormatMessage?.replaceAll(":colors:", message) : undefined;
        setAlertMessage(validProps || alertMessage);
    };

    const setColorPickerHidden = useCallback(
        (hide: boolean): void => {
            if (!disabled && mode !== "inline") {
                setHidden(hide);
            }
        },
        [disabled, mode]
    );
    const onColorValueUpdate = useCallback(
        (color: ColorState): void => {
            if (color && !disabled) {
                const finalColor = parseColor(color, format);
                submitColor(finalColor);
            }
        },
        [disabled, format]
    );

    const renderInput = (): ReactElement => {
        const colorValue = currentColor || color;
        return (
            <Input
                color={colorValue}
                disabled={disabled}
                onKeyUp={setColorPickerHidden}
                onChange={({ target }) => {
                    submitColor(target.value);
                    props.onChange();
                }}
            >
                {renderButton()}
            </Input>
        );
    };
    const onChangeComplete = (color: ColorState): void => {
        if (currentColor !== parseColor(color, format)) {
            props.onChange();
        }
    };
    const renderButton = (): ReactElement => {
        return <Button mode={mode} disabled={disabled} onClick={() => setColorPickerHidden(!hidden)} color={color} />;
    };
    const renderColorPicker = (): ReactElement => {
        const isTriangleAvailable = type === "github" || type === "block" || type === "twitter";
        const supportPopover = mode !== "inline" && type !== "hue" && type !== "slider";
        const colors = defaultColors.map(color => color.color);
        const config = {
            color: color || defaultColor[format],
            colors: defaultColors.length > 0 && type !== "swatches" ? colors : undefined,
            onChangeComplete,
            presetColors: defaultColors.length > 0 ? colors : undefined,
            disableAlpha: format !== "rgba",
            disabled,
            displayColorPicker: hidden,
            onChange: onColorValueUpdate,
            close: () => setColorPickerHidden(true),
            ...(isTriangleAvailable && { triangle: "hide" })
        };
        return (
            <div
                className={classNames({
                    "widget-color-picker-popover": supportPopover,
                    "widget-color-picker-no-popover": mode !== "inline"
                })}
            >
                {mode !== "inline" ? (
                    <div className={"widget-color-picker-cover"} onClick={() => setColorPickerHidden(true)} />
                ) : null}
                {disabled ? <div className={"widget-color-picker-overlay"} /> : null}
                {createElement(colorElement, { ...config })}
            </div>
        );
    };

    useEffect(() => {
        if (color) {
            validateColor(color);
        }
    }, [color]);
    return (
        <div
            className={classNames("widget-color-picker widget-color-picker-picker", {
                "widget-color-picker-disabled": disabled,
                "has-error": !!alertMessage
            })}
        >
            {mode === "input" && renderInput()}
            {mode === "popover" && renderButton()}
            {hidden || (mode !== "inline" && disabled) ? null : renderColorPicker()}
            {alertMessage ? (
                <Alert bootstrapStyle="danger" className={"widget-color-picker-alert has-error"}>
                    {alertMessage}
                </Alert>
            ) : null}
        </div>
    );
};
