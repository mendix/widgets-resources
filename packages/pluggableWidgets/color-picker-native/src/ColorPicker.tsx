import { flattenStyles } from "@native-mobile-resources/util-widgets";
import { ValueStatus } from "mendix";
import { Component, createElement, ReactElement, ReactNode } from "react";
import { View, ViewStyle } from "react-native";

import { HueGradient, LightnessGradient, SaturationGradient } from "react-native-color";
import tinycolor from "tinycolor2";
import { ColorPickerProps } from "../typings/ColorPickerProps";
import { PickerSlider } from "./components/PickerSlider";
import { AlphaGradient } from "./components/AlphaGradient";
import { DisabledHueGradient } from "./components/DisabledHueGradient";
import { ColorPickerStyle, defaultColorPickerStyle } from "./ui/Styles";
import HSLA = tinycolor.ColorFormats.HSLA;
import { executeAction } from "@widgets-resources/piw-utils";

interface State {
    color?: HSLA;
}

export type Props = ColorPickerProps<ColorPickerStyle>;

export class ColorPicker extends Component<Props, State> {
    private readonly onChangeHueHandler = this.onChangeHue.bind(this);
    private readonly onChangeSaturationHandler = this.onChangeSaturation.bind(this);
    private readonly onChangeLightnessHandler = this.onChangeLightness.bind(this);
    private readonly onChangeAlphaHandler = this.onChangeAlpha.bind(this);
    private readonly onChangeCompleteHandler = this.onChangeComplete.bind(this);
    private readonly styles = flattenStyles(defaultColorPickerStyle, this.props.style);
    private readonly defaultSteps = 80;
    readonly state: State = {
        color: undefined
    };

    render(): ReactNode {
        if (!this.props.color || this.props.color.status !== ValueStatus.Available || !this.props.color.value) {
            return null;
        }

        const colorHex = this.state.color ? this.getColor() : this.props.color.value;
        const color = tinycolor(colorHex).toHsl();

        return (
            <View style={this.styles.container} testID={`${this.props.name}`}>
                {this.renderPreview(color)}
                {this.renderHue(color)}
                {this.renderSaturation(color)}
                {this.renderLightness(color)}
                {this.props.format !== "hex" && this.renderAlpha(color)}
            </View>
        );
    }

    componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<State>): void {
        if (this.props.color.value !== prevProps.color.value && this.state.color === prevState.color) {
            this.setState({ color: undefined });
        }
    }

    private onChangeHue(value: number): void {
        const color = tinycolor(this.props.color.value).toHsl();
        color.h = value;
        this.setState({ color });
    }

    private onChangeSaturation(value: number): void {
        const color = tinycolor(this.props.color.value).toHsl();
        color.s = value;
        this.setState({ color });
    }

    private onChangeLightness(value: number): void {
        const color = tinycolor(this.props.color.value).toHsl();
        color.l = value;
        this.setState({ color });
    }

    private onChangeAlpha(value: number): void {
        const color = tinycolor(this.props.color.value).toHsl();
        color.a = value;
        this.setState({ color });
    }

    private onChangeComplete(): void {
        if (this.state.color && this.props.color.value !== this.getColor()) {
            this.props.color.setValue(this.getColor());
        }

        executeAction(this.props.onChange);
    }

    private getColor(): string {
        const color = tinycolor(this.state.color);
        switch (this.props.format) {
            case "hex":
                return color.toHexString();
            case "hsl":
                return color.toHslString();
            case "hsv":
                return color.toHsvString();
            case "rgb":
                return color.toRgbString();
        }
    }

    private renderPreview(color: HSLA): ReactNode {
        return (
            this.props.showPreview && (
                <View
                    testID={`${this.props.name}$preview`}
                    style={[this.styles.preview, { backgroundColor: tinycolor(color).toHslString() }]}
                />
            )
        );
    }

    private renderHue(color: HSLA): ReactElement {
        const component = this.props.color.readOnly ? (
            <DisabledHueGradient color={color} />
        ) : (
            <HueGradient gradientSteps={this.defaultSteps} />
        );
        return (
            <PickerSlider
                testID={`${this.props.name}$hue`}
                value={color.h}
                onValueChange={this.onChangeHueHandler}
                onValueChangeComplete={this.onChangeCompleteHandler}
                step={1}
                maximumValue={359}
                thumbTintColor={tinycolor(color).toHslString()}
                thumbStyle={this.getThumbStyle(color)}
                disabled={this.props.color.readOnly}
            >
                {component}
            </PickerSlider>
        );
    }

    private renderSaturation(color: HSLA): ReactNode {
        if (!this.props.showSaturation || this.props.color.readOnly) {
            return null;
        }
        return (
            <PickerSlider
                testID={`${this.props.name}$saturation`}
                value={color.s}
                onValueChange={this.onChangeSaturationHandler}
                onValueChangeComplete={this.onChangeCompleteHandler}
                step={0.01}
                thumbTintColor={tinycolor(color).toHslString()}
                thumbStyle={this.getThumbStyle(color)}
                disabled={this.props.color.readOnly}
            >
                <SaturationGradient color={color} gradientSteps={this.defaultSteps} />
            </PickerSlider>
        );
    }

    private renderLightness(color: HSLA): ReactNode {
        if (!this.props.showLightness || this.props.color.readOnly) {
            return null;
        }
        return (
            <PickerSlider
                testID={`${this.props.name}$lightness`}
                value={color.l}
                onValueChange={this.onChangeLightnessHandler}
                onValueChangeComplete={this.onChangeCompleteHandler}
                step={0.01}
                thumbTintColor={tinycolor(color).toHslString()}
                thumbStyle={this.getThumbStyle(color)}
                disabled={this.props.color.readOnly}
            >
                <LightnessGradient color={color} gradientSteps={this.defaultSteps} />
            </PickerSlider>
        );
    }

    private renderAlpha(color: HSLA): ReactNode {
        if (!this.props.showAlpha || this.props.color.readOnly) {
            return null;
        }
        return (
            <PickerSlider
                testID={`${this.props.name}$alpha`}
                value={color.a}
                onValueChange={this.onChangeAlphaHandler}
                onValueChangeComplete={this.onChangeCompleteHandler}
                step={0.01}
                thumbTintColor={tinycolor(color).toHslString()}
                thumbStyle={this.getThumbStyle(color)}
                disabled={this.props.color.readOnly}
            >
                <AlphaGradient color={color} gradientSteps={this.defaultSteps} />
            </PickerSlider>
        );
    }

    private readonly getThumbStyle = (color: HSLA): ViewStyle => ({
        borderWidth: 1,
        borderColor: tinycolor(color).toHexString()
    });
}
