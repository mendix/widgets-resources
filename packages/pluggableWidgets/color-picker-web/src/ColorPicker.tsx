import { createElement, ReactNode, useCallback } from "react";
import { executeAction } from "@mendix/piw-utils-internal";
import { ColorPicker as ColorPickerComponent } from "./components/ColorPicker";
import { ColorPickerContainerProps } from "../typings/ColorPickerProps";
import "./ui/ColorPicker.scss";

export function ColorPicker(props: ColorPickerContainerProps): ReactNode {
    const { name, mode, tabIndex, type, onChange, colorAttribute, defaultColors, format, id, invalidFormatMessage } =
        props;
    const onChangeFn = useCallback(() => executeAction(onChange), [onChange]);
    const onColorChange = useCallback(value => props.colorAttribute.setValue(value), [props.colorAttribute]);
    return (
        <ColorPickerComponent
            id={id}
            name={name}
            color={colorAttribute.value}
            onChange={onChangeFn}
            onColorChange={onColorChange}
            tabIndex={tabIndex}
            mode={mode}
            type={type}
            defaultColors={defaultColors}
            format={format}
            disabled={colorAttribute.readOnly}
            invalidFormatMessage={invalidFormatMessage?.value}
        />
    );
}
