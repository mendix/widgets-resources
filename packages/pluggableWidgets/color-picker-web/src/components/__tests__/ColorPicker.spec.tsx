import { createElement } from "react";
import { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import { ColorPicker, ColorPickerProps } from "../ColorPicker";

describe("ColorPicker", () => {
    const renderColorPicker = (props: ColorPickerProps): ShallowWrapper<ColorPickerProps, any> =>
        shallow(<ColorPicker {...props} />);
    const fullRenderColorPicker = (props: ColorPickerProps): ReactWrapper<ColorPickerProps, any> =>
        mount(<ColorPicker {...props} />);
    const colorPickerProps: ColorPickerProps = {
        color: "#000000",
        disabled: false,
        defaultColors: [],
        format: "hex",
        mode: "popover",
        type: "sketch",
        onChange: jest.fn(),
        onColorChange: jest.fn(),
        id: "color-picker",
        name: "color picker"
    };

    it("renders the structure correctly", () => {
        const colorPickerComponent = renderColorPicker(colorPickerProps);

        expect(colorPickerComponent).toMatchSnapshot();
    });

    it("that is disabled renders with the structure", () => {
        const colorPickerComponent = renderColorPicker({ ...colorPickerProps, disabled: true });
        expect(colorPickerComponent).toMatchSnapshot();
    });

    it("renders picker with pre-defined default colors", () => {
        const colorPickerComponent = renderColorPicker(colorPickerProps);
        colorPickerComponent.setProps({
            mode: "inline",
            defaultColors: [{ color: "#2CCCE4" }, { color: "#555555" }] as any
        });

        expect(colorPickerComponent).toMatchSnapshot();
    });

    it("should handle on change event", () => {
        const colorPickerComponent = fullRenderColorPicker({
            ...colorPickerProps,
            type: "block",
            mode: "inline",
            defaultColors: [{ color: "#F47373" }]
        });
        const colorElement = colorPickerComponent.find("[title='#F47373']");
        colorElement.simulate("click");
        expect(colorPickerProps.onColorChange).toHaveBeenCalled();
    });

    describe("renders a picker of type", () => {
        it("sketch", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "sketch" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".sketch-picker")).toHaveLength(1);
        });

        it("chrome", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "chrome" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".chrome-picker")).toHaveLength(1);
        });

        it("block", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "block" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");

            expect(colorPickerComponent.find(".block-picker")).toHaveLength(1);
        });

        it("github", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "github" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".github-picker")).toHaveLength(1);
        });

        it("twitter", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "twitter" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".twitter-picker")).toHaveLength(1);
        });

        it("circle", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "circle" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".circle-picker")).toHaveLength(1);
        });

        it("hue", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "hue" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".hue-picker")).toHaveLength(1);
        });

        it("slider", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "slider" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".slider-picker")).toHaveLength(1);
        });

        it("compact", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "compact" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".compact-picker")).toHaveLength(1);
        });

        it("material", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "material" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".material-picker")).toHaveLength(1);
        });

        it("swatches", () => {
            const colorPickerComponent = fullRenderColorPicker({ ...colorPickerProps, type: "swatches" });
            const colorElement = colorPickerComponent.find("button");
            colorElement.simulate("click");
            expect(colorPickerComponent.find(".swatches-picker")).toHaveLength(1);
        });
    });

    describe("with a mode as", () => {
        it("popover or input renders with the structure", () => {
            const colorPickerComponent = renderColorPicker({ ...colorPickerProps, mode: "popover" });

            expect(colorPickerComponent).toMatchSnapshot();
        });

        it("inline renders with the structure", () => {
            const colorPickerComponent = renderColorPicker({ ...colorPickerProps, mode: "inline" });
            expect(colorPickerComponent).toMatchSnapshot();
        });
    });
});
