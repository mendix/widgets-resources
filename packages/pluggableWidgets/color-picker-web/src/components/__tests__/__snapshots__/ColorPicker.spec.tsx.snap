// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker renders picker with pre-defined default colors 1`] = `
<div
  className="widget-color-picker widget-color-picker-picker"
/>
`;

exports[`ColorPicker renders the structure correctly 1`] = `
<div
  className="widget-color-picker widget-color-picker-picker"
>
  <Button
    color="#000000"
    disabled={false}
    mode="popover"
    onClick={[Function]}
  />
</div>
`;

exports[`ColorPicker that is disabled renders with the structure 1`] = `
<div
  className="widget-color-picker widget-color-picker-picker widget-color-picker-disabled"
>
  <Button
    color="#000000"
    disabled={true}
    mode="popover"
    onClick={[Function]}
  />
</div>
`;

exports[`ColorPicker with a mode as inline renders with the structure 1`] = `
<div
  className="widget-color-picker widget-color-picker-picker"
>
  <div
    className=""
  >
    <ColorPicker
      close={[Function]}
      color="#000000"
      disableAlpha={true}
      disabled={false}
      displayColorPicker={false}
      onChange={[Function]}
      onChangeComplete={[Function]}
      presetColors={
        Array [
          "#D0021B",
          "#F5A623",
          "#F8E71C",
          "#8B572A",
          "#7ED321",
          "#417505",
          "#BD10E0",
          "#9013FE",
          "#4A90E2",
          "#50E3C2",
          "#B8E986",
          "#000000",
          "#4A4A4A",
          "#9B9B9B",
          "#FFFFFF",
        ]
      }
      styles={Object {}}
      width={200}
    />
  </div>
</div>
`;

exports[`ColorPicker with a mode as popover or input renders with the structure 1`] = `
<div
  className="widget-color-picker widget-color-picker-picker"
>
  <Button
    color="#000000"
    disabled={false}
    mode="popover"
    onClick={[Function]}
  />
</div>
`;
